src/model/openthaigpt_pretraining_model/lightning/utils.py

```diff
@@ -0,0 +1,219 @@
+            while len(buffer) > self.max_tokens:
+                yield torch.tensor(buffer[: self.max_tokens])
+                buffer = buffer[self.max_tokens :]
+
+
+def seed_everything(seed):
+    torch.manual_seed(seed)
+    torch.cuda.manual_seed(seed)
+    torch.cuda.manual_seed_all(seed)
+    torch.backends.cudnn.deterministic = True
+    torch.backends.cudnn.benchmark = False
+    np.random.seed(seed)
+    random.seed(seed)
+
+
+class Trainer:
+    def __init__(
+        self,
+        accelerator: Union[str, Accelerator] = "auto",
+        strategy: Union[str, Strategy] = "auto",
+        devices: Union[List[int], str, int] = "auto",
+        precision: Union[str, int] = "32-true",
+        seed: int = 42,
+        batch_size: int = 8,
+        grad: int = 4,
+        context_length: int = 256,
+        model_name: str = "llama",
+        optimizer: str = "adamw",
+        weight_decay: float = 1e-2,
+        lr: float = 1e-4,
+    ):
+        self.max_tokens = context_length
+        self.step = 0
+        self.seed = seed
+        self.grad = grad
+        self.fabric = L.Fabric(
+            accelerator=accelerator,
+            strategy=strategy,
+            devices=devices,
+            precision=precision,
+        )
+        self.fabric.launch()
+        if model_name == "llama":
+            model_name = LLAMA_MODEL  # for tokenizer
+            # cfg = ModelArgs(
+            #     dim=512,
+            #     n_layers=8,
+            #     n_heads=8,
+            #     vocab_size=32000,
+            #     multiple_of=256,
+            #     norm_eps=1e-5,
+            #     max_batch_size=32,
+            #     max_seq_len=2048,
+            #     attention_mode=ORIGIN_ATTENTION_MODE,
+            # )
+            # self.model = model = Transformer(cfg)
+            cfg = LlamaConfig(
+                vocab_size=32000,
+                hidden_size=1024,
+                num_hidden_layers=8,
+                num_attention_heads=8,
+                hidden_act="silu",
+                max_position_embeddings=2048,  # sequence length
+            )
+            self.model = model = LlamaForCausalLM(cfg)
+        elif model_name == "gptj":
+            model_name = GPTJ_MODEL  # for tokenizer
+            cfg = GPTJConfig(
+                vocab_size=50400,
+                n_positions=2048,
+                n_embd=1536,
+                n_layer=12,
+                n_head=8,
+                rotary_dim=64,
+                n_inner=None,
+                activation_function="gelu_new",
+                resid_pdrop=0.1,
+                embd_pdrop=0.1,
+                attn_pdrop=0.1,
+                layer_norm_epsilon=1e-5,
+                initializer_range=0.02,
+                use_cache=True,
+                bos_token_id=50256,
+                eos_token_id=50256,
+                tie_word_embeddings=False,
+            )
+            self.model = model = GPTJForCausalLM(cfg)
```

Comment:
Please use our own implementation of GPTJ

Reply:
If I follow our implementation, do you mean to change the following line instead?
```  GPTJModel.from_pretrained(pretrained_name).to(device)
    if use_xformers:
        if device == "cpu":
            GPTJAttention._attn = _attn_xformers_cpu
        else:
            GPTJAttention._attn = _attn_xformers 
    else:
        print("Use original")
        GPTJAttention._attn = _attn_orig ```